src/eir.rs

```diff
@@ -0,0 +1,250 @@
+                if data.remaining() % 2 != 0 {
+                    return Err(EIRError::UnexpectedDataLength { len:data.remaining() });
+                }
+                while data.has_remaining() {
+                    eir.uuid16.push(data.get_u16_le());
+                }
+            },
+            Some(EIRDataTypes::UUID32Incomplete)|Some(EIRDataTypes::UUID32Complete) => {
+                if data.remaining() % 4 != 0 {
+                    return Err(EIRError::UnexpectedDataLength { len:data.remaining() });
+                }
+                while data.has_remaining() {
+                    eir.uuid32.push(data.get_u32_le());
+                }
+            },
+            Some(EIRDataTypes::UUID128Incomplete)|Some(EIRDataTypes::UUID128Complete) => {
+                if data.remaining() % 16 != 0 {
+                    return Err(EIRError::UnexpectedDataLength { len:data.remaining() });                    
+                }
+                while data.has_remaining() {
+                    eir.uuid128.push(data.get_u128_le());
+                }
+            },
+            Some(EIRDataTypes::NameShort) => {
+                if eir.name.is_some() {
+                    return Err(EIRError::RepeatedName);
+                }
+                eir.name = Some(EIRName::short_name(String::from_utf8_lossy(data.bytes()).to_string()));
+            },
+            Some(EIRDataTypes::NameComplete) => {
+                if eir.name.is_some() {
+                    return Err(EIRError::RepeatedName);
+                }
+                eir.name = Some(EIRName::complete_name(String::from_utf8_lossy(data.bytes()).to_string()));
+            },
+            Some(EIRDataTypes::TxPowerLevel) => {
+                eir.tx_power_level.push(data.get_i8());
+            },
+            Some(EIRDataTypes::URI) => {
+                let uri_scheme = data.get_u8();
+                if uri_scheme == 0x01 {
+                    let uri = String::from_utf8(data.bytes().to_vec());
+                    if uri.is_err() {
+                        return Err(EIRError::InvalidURI);
+                    }
+                    eir.uri.push(uri.unwrap());
+                } else {
+                    // TODO: URI scheme translation. Skip for now.
+                }
+            },
+            Some(EIRDataTypes::ManufacturerSpecificData) => {
+                if data.remaining() < 2 {
+                    return Err(EIRError::UnexpectedDataLength { len:data.remaining() });
+                }
+                eir.manufacturer_specific_data.push(
+                    ManufacturerSpecificData {
+                        company_identifier_code: data.get_u16_le(),
+                        data: Bytes::copy_from_slice(data.bytes()),
+                    }
+                );
+            },
+            _ => {
+                // Skip unknown data
+            },
+        }
+        data.advance(data.remaining());
+        buf = data.into_inner();
```

Comment:
I had to restore buf after moving it earlier. Otherwise I get the following error.

```
error[E0382]: borrow of moved value: `buf`
   --> src/eir.rs:105:11
    |
102 | pub fn parse_eir<T: Buf>(mut buf: T) -> Result<EIR, EIRError> {
    |                          ------- move occurs because `buf` has type `T`, which does not implement the `Copy` trait
...
105 |     while buf.has_remaining() {
    |           ^^^ value borrowed here after move
...
121 |         let mut data = buf.take((len - 1).into());
    |                        --- value moved here, in previous iteration of loop
    |
help: consider further restricting this bound
    |
102 | pub fn parse_eir<T: Buf + Copy>(mut buf: T) -> Result<EIR, EIRError> {
    |                         ^^^^^^

error: aborting due to previous error
```

I'm fairly new to Rust. So there might be an obvious way of doing this I'm missing.

Reply:
It's not really clear why you need to call `buf.take()`. Avoiding `.take()` would solve the problem, as `.take()` is the call that moves `buf`. You can still do all of the bounds checking that you do later in the loop using just `len`, and if you need to create a contiguous series of bytes then you can do `buf.copy_to_bytes()`.

Note that [`copy_to_bytes()`](https://docs.rs/bytes/0.6.0/bytes/trait.Buf.html#method.copy_to_bytes) was added in `bytes` v0.6; you will need to rebase your branch against `master`.